homeassistant/components/restore_state.py

```diff
@@ -0,0 +1,86 @@
+"""
+Support for restoring entity states on startup.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, ATTR_RESTORED_STATE)
+from homeassistant.components.history import (
+    CONF_DOMAINS, CONF_EXCLUDE, CONF_ENTITIES, CONF_INCLUDE,
+    Filters, get_states, last_recorder_run)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+import homeassistant.util.dt as dt_util
+
+DOMAIN = 'restore_state'
+
+DEPENDENCES = ['recorder']
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: cv.FILTER_SCHEMA,
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ALWAYS_OVERWRITE = (
```

Comment:
In a way, overwriting works well. Although the Internal state differs, the HASS state is used in automation, the UI, etc. You also automatically drop the "restored" attribute once it changes. This does not work for polled entities though!

@balloob the new hook on Entity component sounds like good solution, we can then gracefully catch NotImplemnted exceptions as well, so ignore components not supporting this hook. Must still wrap my mind around why not simply calling the new hook on HOMEASSISTANT_START (the drawback would probably be multiple state changes at start)

The drawback of restoring it into the states is that we might end up with "ghost" states from configuration past. Although these ghosts should disappear on next startup since no state was changed.

@ronvl You can limit domains or entities for restoration today through filters, for example `exclude: {domain: [mqtt]}}` or `include: {domain: [input_select]}`